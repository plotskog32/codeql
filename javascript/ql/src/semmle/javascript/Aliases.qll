/**
 * Provides aliases for commonly used classes that have different names
 * in the QL libraries for other languages.
 */

import javascript

class AndBitwiseExpr = BitAndExpr;

class AndLogicalExpr = LogAndExpr;

class ArrayAccess = IndexExpr;

class AssignOp = CompoundAssignExpr;

class Block = BlockStmt;

class BoolLiteral = BooleanLiteral;

class CaseStmt = Case;

class ComparisonOperation = Comparison;

class DoStmt = DoWhileStmt;

class EqualityOperation = EqualityTest;

class FieldAccess = DotExpr;

class InstanceOfExpr = InstanceofExpr;

class LabelStmt = LabeledStmt;

class LogicalAndExpr = LogAndExpr;

class LogicalNotExpr = LogNotExpr;

class LogicalOrExpr = LogOrExpr;

class Loop = LoopStmt;

class MultilineComment = BlockComment;

class OrBitwiseExpr = BitOrExpr;

class OrLogicalExpr = LogOrExpr;

class ParenthesisExpr = ParExpr;

class ParenthesizedExpr = ParExpr;

class RelationalOperation = RelationalComparison;

class RemExpr = ModExpr;

class SingleLineComment = LineComment;

class SuperAccess = SuperExpr;

class SwitchCase = Case;

class ThisAccess = ThisExpr;

class VariableAccess = VarAccess;

class XorBitwiseExpr = XOrExpr;

// Aliases for deprecated predicates from the dbscheme
 
/**
 * Alias for the predicate `is_externs` defined in the .dbscheme.
 * Use `TopLevel#isExterns()` instead.
 */ 
deprecated predicate isExterns(TopLevel toplevel) { is_externs(toplevel) }
/**
 * Alias for the predicate `is_module` defined in the .dbscheme.
 * Use the `Module` class in `Module.qll` instead.
 */ 
deprecated predicate isModule(TopLevel toplevel) { is_module(toplevel) }
/**
 * Alias for the predicate `is_nodejs` defined in the .dbscheme.
 * Use `NodeModule` from `NodeJS.qll` instead.
 */ 
deprecated predicate isNodejs(TopLevel toplevel) { is_nodejs(toplevel) }
/**
 * Alias for the predicate `is_es2015_module` defined in the .dbscheme.
 * Use `ES2015Module` from `ES2015Modules.qll` instead.
 */ 
deprecated predicate isES2015Module(TopLevel toplevel) { is_es2015_module(toplevel) }
/**
 * Alias for the predicate `is_closure_module` defined in the .dbscheme.
 */ 
deprecated predicate isClosureModule(TopLevel toplevel) { is_closure_module(toplevel) }
/**
 * Alias for the predicate `stmt_containers` defined in the .dbscheme.
 * Use `ASTNode#getContainer()` instead.
 */ 
deprecated predicate stmtContainers(Stmt stmt, StmtContainer container) { stmt_containers(stmt, container) }
/**
 * Alias for the predicate `jump_targets` defined in the .dbscheme.
 * Use `JumpStmt#getTarget()` instead.
 */ 
deprecated predicate jumpTargets(Stmt jump, Stmt target) { jump_targets(jump, target) }
/**
 * Alias for the predicate `is_instantiated` defined in the .dbscheme.
 * Use `NamespaceDeclaration#isInstantiated() instead.`
 */ 
deprecated predicate isInstantiated(NamespaceDeclaration decl) { is_instantiated(decl) }
/**
 * Alias for the predicate `has_declare_keyword` defined in the .dbscheme.
 */ 
deprecated predicate hasDeclareKeyword(ASTNode stmt) { has_declare_keyword(stmt) }
/**
 * Alias for the predicate `is_for_await_of` defined in the .dbscheme.
 * Use `ForOfStmt#isAwait()` instead.
 */ 
deprecated predicate isForAwaitOf(ForOfStmt forof) { is_for_await_of(forof) }
/**
 * Alias for the predicate `enclosing_stmt` defined in the .dbscheme.
 * Use `ExprOrType#getEnclosingStmt` instead.
 */ 
deprecated predicate enclosingStmt(ExprOrType expr, Stmt stmt) { enclosing_stmt(expr, stmt) }
/**
 * Alias for the predicate `expr_containers` defined in the .dbscheme.
 * Use `ASTNode#getContainer()` instead.
 */ 
deprecated predicate exprContainers(ExprOrType expr, StmtContainer container) { expr_containers(expr, container) }
/**
 * Alias for the predicate `array_size` defined in the .dbscheme.
 * Use `ArrayExpr#getSize()` instead.
 */ 
deprecated predicate arraySize(Expr ae, int sz) { array_size(ae, sz) }
/**
 * Alias for the predicate `is_delegating` defined in the .dbscheme.
 * Use `YieldExpr#isDelegating()` instead.
 */ 
deprecated predicate isDelegating(YieldExpr yield) { is_delegating(yield) }
/**
 * Alias for the predicate `is_arguments_object` defined in the .dbscheme.
 * Use the `ArgumentsVariable` class instead.
 */ 
deprecated predicate isArgumentsObject(Variable id) { is_arguments_object(id) }
