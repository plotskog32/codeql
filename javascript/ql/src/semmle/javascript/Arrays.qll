import javascript
private import semmle.javascript.dataflow.InferredTypes

/**
 * Classes and predicates for modelling TaintTracking steps for arrays.
 */ 
module ArrayTaintTracking {
  /**
   * A taint propagating data flow edge caused by the builtin array functions.
   */
  private class ArrayFunctionTaintStep extends TaintTracking::AdditionalTaintStep {
    DataFlow::CallNode call;

    ArrayFunctionTaintStep() { this = call }

    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
      arrayFunctionTaintStep(pred, succ, call)
    }
  }

  /**
   * A taint propagating data flow edge from `pred` to `succ` caused by a call `call` to a builtin array functions.
   */
  predicate arrayFunctionTaintStep(DataFlow::Node pred, DataFlow::Node succ, DataFlow::CallNode call) {
    // `array.map(function (elt, i, ary) { ... })`: if `array` is tainted, then so are
    // `elt` and `ary`; similar for `forEach`
    exists(string name, Function f, int i |
      (name = "map" or name = "forEach") and
      (i = 0 or i = 2) and
      call.getArgument(0).analyze().getAValue().(AbstractFunction).getFunction() = f and
      call.(DataFlow::MethodCallNode).getMethodName() = name and
      pred = call.getReceiver() and
      succ = DataFlow::parameterNode(f.getParameter(i))
    )
    or
    // `array.map` with tainted return value in callback
    exists(DataFlow::FunctionNode f |
      call.(DataFlow::MethodCallNode).getMethodName() = "map" and
      call.getArgument(0) = f and // Require the argument to be a closure to avoid spurious call/return flow
      pred = f.getAReturn() and
      succ = call
    )
    or
    // `array.push(e)`, `array.unshift(e)`: if `e` is tainted, then so is `array`.
    exists(string name |
      name = "push" or
      name = "unshift"
    |
      pred = call.getAnArgument() and
      succ.(DataFlow::SourceNode).getAMethodCall(name) = call
    )
    or
    // `array.push(...e)`, `array.unshift(...e)`: if `e` is tainted, then so is `array`.
    exists(string name |
      name = "push" or
      name = "unshift"
    |
      pred = call.getASpreadArgument() and
      // Make sure we handle reflective calls
      succ = call.getReceiver().getALocalSource() and
      call.getCalleeName() = name
    )
    or
    // `array.splice(i, del, e)`: if `e` is tainted, then so is `array`.
    exists(string name | name = "splice" |
      pred = call.getArgument(2) and
      succ.(DataFlow::SourceNode).getAMethodCall(name) = call
    )
    or
    // `e = array.pop()`, `e = array.shift()`, or similar: if `array` is tainted, then so is `e`.
    exists(string name |
      name = "pop" or
      name = "shift" or
      name = "slice" or
      name = "splice"
    |
      call.(DataFlow::MethodCallNode).calls(pred, name) and
      succ = call
    )
    or
    // `e = Array.from(x)`: if `x` is tainted, then so is `e`.
    call = DataFlow::globalVarRef("Array").getAPropertyRead("from").getACall() and
    pred = call.getAnArgument() and
    succ = call
    or
    // `e = arr1.concat(arr2, arr3)`: if any of the `arr` is tainted, then so is `e`.
    call.(DataFlow::MethodCallNode).calls(pred, "concat") and
    succ = call
    or
    call.(DataFlow::MethodCallNode).getMethodName() = "concat" and
    succ = call and
    pred = call.getAnArgument()
  }
}
