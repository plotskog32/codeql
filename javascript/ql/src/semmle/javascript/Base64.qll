/**
 * Provides classes and predicates for working with base64 encoders and decoders.
 */

import javascript

module Base64 {
  /** A call to a base64 encoder. */
  class Encode extends DataFlow::Node {
    Encode::Range encode;

    Encode() { this = encode }

    /** Gets the input passed to the encoder. */
    DataFlow::Node getInput() { result = encode.getInput() }

    /** Gets the base64-encoded output of the encoder. */
    DataFlow::Node getOutput() { result = encode.getOutput() }
  }

  module Encode {
    /**
     * A data flow node that should be considered a call to a base64 encoder.
     *
     * New base64 encoding functions can be supported by extending this class.
     */
    abstract class Range extends DataFlow::Node {
      /** Gets the input passed to the encoder. */
      abstract DataFlow::Node getInput();

      /** Gets the base64-encoded output of the encoder. */
      abstract DataFlow::Node getOutput();
    }
  }

  /** A call to a base64 decoder. */
  class Decode extends DataFlow::Node {
    Decode::Range encode;

    Decode() { this = encode }

    /** Gets the base64-encoded input passed to the decoder. */
    DataFlow::Node getInput() { result = encode.getInput() }

    /** Gets the output of the decoder. */
    DataFlow::Node getOutput() { result = encode.getOutput() }
  }

  module Decode {
    /**
     * A data flow node that should be considered a call to a base64 decoder.
     *
     * New base64 decoding functions can be supported by extending this class.
     */
    abstract class Range extends DataFlow::Node {
      /** Gets the base64-encoded input passed to the decoder. */
      abstract DataFlow::Node getInput();

      /** Gets the output of the decoder. */
      abstract DataFlow::Node getOutput();
    }
  }

  /**
   * A base64 decoding step, viewed as a taint-propagating data flow edge.
   *
   * Note that we currently do not model base64 encoding as a taint-propagating data flow edge
   * to avoid false positives.
   */
  private class Base64DecodingStep extends TaintTracking::AdditionalTaintStep {
    Decode dec;

    Base64DecodingStep() { this = dec }

    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
      pred = dec.getInput() and
      succ = dec.getOutput()
    }
  }
}

/** A call to `btoa`. */
private class Btoa extends Base64::Encode::Range, DataFlow::CallNode {
  Btoa() { this = DataFlow::globalVarRef("btoa").getACall() }

  override DataFlow::Node getInput() { result = getArgument(0) }

  override DataFlow::Node getOutput() { result = this }
}

/** A call to `atob`. */
private class Atob extends Base64::Decode::Range, DataFlow::CallNode {
  Atob() { this = DataFlow::globalVarRef("atob").getACall() }

  override DataFlow::Node getInput() { result = getArgument(0) }

  override DataFlow::Node getOutput() { result = this }
}

/** A call to `Buffer.from` with encoding `base64`. */
private class Buffer_from extends Base64::Encode::Range, DataFlow::CallNode {
  Buffer_from() {
    this = DataFlow::globalVarRef("Buffer").getAMemberCall("from") and
    getArgument(1).mayHaveStringValue("base64")
  }

  override DataFlow::Node getInput() { result = getArgument(0) }

  override DataFlow::Node getOutput() { result = this }
}

/**
 * A call to `Buffer.prototype.toString` with encoding `base64`, approximated by
 * looking for calls to `toString` where the first argument is the string `"base64"`.
 */
private class Buffer_toString extends Base64::Decode::Range, DataFlow::MethodCallNode {
  Buffer_toString() {
    getMethodName() = "toString" and
    getArgument(0).mayHaveStringValue("base64")
  }

  override DataFlow::Node getInput() { result = getReceiver() }

  override DataFlow::Node getOutput() { result = this }
}

/**
 * A call to a base64 encoding function from one of the npm packages
 * `base-64`, `js-base64`, `Base64`, or `base64-js`.
 */
private class NpmBase64Encode extends Base64::Encode::Range, DataFlow::CallNode {
  NpmBase64Encode() {
    exists(string mod, string meth |
      mod = "base-64" and meth = "encode"
      or
      mod = "Base64" and meth = "btoa"
      or
      mod = "base64-js" and meth = "toByteArray"
    |
      this = DataFlow::moduleMember(mod, meth).getACall()
    )
    or
    exists(string meth | meth = "encode" or meth = "encodeURI" |
      this = DataFlow::moduleMember("js-base64", "Base64").getAMemberCall(meth)
    )
  }

  override DataFlow::Node getInput() { result = getArgument(0) }

  override DataFlow::Node getOutput() { result = this }
}

/**
 * A call to a base64 decoding function from one of the npm packages
 * `base-64`, `js-base64`, `Base64`, or `base64-js`.
 */
private class NpmBase64Decode extends Base64::Decode::Range, DataFlow::CallNode {
  NpmBase64Decode() {
    exists(string mod, string meth |
      mod = "base-64" and meth = "decode"
      or
      mod = "Base64" and meth = "atob"
      or
      mod = "base64-js" and meth = "fromByteArray"
    |
      this = DataFlow::moduleMember(mod, meth).getACall()
    )
    or
    this = DataFlow::moduleMember("js-base64", "Base64").getAMemberCall("decode")
  }

  override DataFlow::Node getInput() { result = getArgument(0) }

  override DataFlow::Node getOutput() { result = this }
}
