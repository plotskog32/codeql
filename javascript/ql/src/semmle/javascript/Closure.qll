/**
 * Provides classes for working with the Closure-Library module system.
 */

import javascript

module Closure {
  /**
   * A call to a function in the `goog` namespace such as `goog.provide` or `goog.load`.
   */
  class GoogFunctionCall extends CallExpr {
    GoogFunctionCall() {
      exists(GlobalVariable gv | gv.getName() = "goog" |
        this.getCallee().(DotExpr).getBase() = gv.getAnAccess()
      )
    }

    /** Gets the name of the invoked function. */
    string getFunctionName() { result = getCallee().(DotExpr).getPropertyName() }
  }

  /**
   * An expression statement consisting of a call to a function
   * in the `goog` namespace.
   */
  class GoogFunctionCallStmt extends ExprStmt {
    GoogFunctionCallStmt() { super.getExpr() instanceof GoogFunctionCall }

    override GoogFunctionCall getExpr() { result = super.getExpr() }

    /** Gets the name of the invoked function. */
    string getFunctionName() { result = getExpr().getFunctionName() }

    /** Gets the `i`th argument to the invoked function. */
    Expr getArgument(int i) { result = getExpr().getArgument(i) }

    /** Gets an argument to the invoked function. */
    Expr getAnArgument() { result = getArgument(_) }
  }

  abstract private class GoogNamespaceRef extends ExprOrStmt { abstract string getNamespaceId(); }

  /**
   * A call to `goog.provide`.
   */
  class GoogProvide extends GoogFunctionCallStmt, GoogNamespaceRef {
    GoogProvide() { getFunctionName() = "provide" }

    /** Gets the identifier of the namespace created by this call. */
    override string getNamespaceId() { result = getArgument(0).getStringValue() }
  }

  /**
   * A call to `goog.require`.
   */
  class GoogRequire extends GoogFunctionCall, GoogNamespaceRef {
    GoogRequire() { getFunctionName() = "require" }

    /** Gets the identifier of the namespace imported by this call. */
    override string getNamespaceId() { result = getArgument(0).getStringValue() }
  }

  private class GoogRequireImport extends GoogRequire, Import {
    /** Gets the module in which this import appears. */
    override Module getEnclosingModule() { result = getTopLevel() }

    /** Gets the (unresolved) path that this import refers to. */
    override PathExpr getImportedPath() { result = getArgument(0) }
  }

  /**
   * A call to `goog.module` or `goog.declareModuleId`.
   */
  class GoogModuleDeclaration extends GoogFunctionCallStmt, GoogNamespaceRef {
    GoogModuleDeclaration() {
      getFunctionName() = "module" or
      getFunctionName() = "declareModuleId"
    }

    /** Gets the identifier of the namespace imported by this call. */
    override string getNamespaceId() { result = getArgument(0).getStringValue() }
  }

  /**
   * A module using the Closure module system, declared using `goog.module()` or `goog.declareModuleId()`.
   */
  class ClosureModule extends Module {
    ClosureModule() { getAChildStmt() instanceof GoogModuleDeclaration }

    /**
     * Gets the call to `goog.module` or `goog.declareModuleId` in this module.
     */
    GoogModuleDeclaration getModuleDeclaration() { result = getAChildStmt() }

    /**
     * Gets the namespace of this module.
     */
    string getNamespaceId() { result = getModuleDeclaration().getNamespaceId() }

    override Module getAnImportedModule() {
      exists(GoogRequireImport imprt |
        imprt.getEnclosingModule() = this and
        result.(ClosureModule).getNamespaceId() = imprt.getNamespaceId()
      )
    }

    /**
     * Gets the top-level `exports` variable in this module, if this module is defined by
     * a `good.module` call.
     *
     * This variable denotes the object exported from this module.
     *
     * Has no result for ES6 modules using `goog.declareModuleId`.
     */
    Variable getExportsVariable() {
      getModuleDeclaration().getFunctionName() = "module" and
      result = getScope().getVariable("exports")
    }

    override predicate exports(string name, ASTNode export) {
      exists(DataFlow::PropWrite write, Expr base |
        write.getAstNode() = export and
        write.writes(base.flow(), name, _) and
        (
          base = getExportsVariable().getAReference()
          or
          base = getExportsVariable().getAnAssignedExpr()
        )
      )
    }
  }

  /**
   * A global Closure script, that is, a toplevel that is executed in the global scope and
   * contains a toplevel call to `goog.provide` or `goog.require`.
   */
  class ClosureScript extends TopLevel {
    ClosureScript() {
      not this instanceof ClosureModule and
      getAChildStmt() instanceof GoogProvide
      or
      getAChildStmt().(ExprStmt).getExpr() instanceof GoogRequire
    }

    /** Gets the identifier of a namespace required by this module. */
    string getARequiredNamespace() {
      result = getAChildStmt().(ExprStmt).getExpr().(GoogRequire).getNamespaceId()
    }

    /** Gets the identifer of a namespace provided by this module. */
    string getAProvidedNamespace() { result = getAChildStmt().(GoogProvide).getNamespaceId() }
  }

  /**
   * Holds if `name` is a closure namespace, including proper namespace prefixes.
   */
  pragma[noinline]
  predicate isLibraryNamespacePath(string name) {
    exists(string namespace | namespace = any(GoogNamespaceRef provide).getNamespaceId() |
      name = namespace.substring(0, namespace.indexOf("."))
      or
      name = namespace
    )
  }

  /**
   * Gets the closure namespace path addressed by the given dataflow node, if any.
   */
  string getLibraryAccessPath(DataFlow::SourceNode node) {
    isLibraryNamespacePath(result) and
    node = DataFlow::globalVarRef(result)
    or
    isLibraryNamespacePath(result) and
    exists(DataFlow::PropRead read | node = read |
      result = getLibraryAccessPath(read.getBase().getALocalSource()) + "." + read.getPropertyName()
    )
    or
    // Associate an access path with the immediate RHS of a store on a closure namespace.
    // This is to support patterns like:
    // foo.bar = { baz() {} }
    exists(DataFlow::PropWrite write |
      node = write.getRhs() and
      result = getWrittenLibraryAccessPath(write)
    )
    or
    result = node.asExpr().(GoogRequire).getNamespaceId()
  }

  /**
   * Gets the closure namespace path written to by the given property write, if any.
   */
  string getWrittenLibraryAccessPath(DataFlow::PropWrite node) {
    result = getLibraryAccessPath(node.getBase().getALocalSource()) + "." + node.getPropertyName()
  }

  /**
   * Gets a dataflow node that refers to the given Closure module.
   */
  DataFlow::SourceNode moduleImport(string moduleName) {
    getLibraryAccessPath(result) = moduleName
  }

  /**
   * Gets a dataflow node that refers to the given member of a Closure module.
   */
  DataFlow::SourceNode moduleMember(string moduleName, string memberName) {
    result = moduleImport(moduleName).getAPropertyRead(memberName)
  }
}
