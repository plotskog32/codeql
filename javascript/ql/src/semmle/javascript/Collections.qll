/**
 * Provides predicates and classes for working with the standard library collection implementations.
 * Currently [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) and
 * [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) are implemented.
 */

import javascript
private import semmle.javascript.dataflow.internal.StepSummary
private import DataFlow::PseudoProperties

/**
 * An `AdditionalFlowStep` used to model a data-flow step related to standard library collections.
 *
 * The `loadStep`/`storeStep`/`loadStoreStep` methods are overloaded such that the new predicates
 * `load`/`store`/`loadStore` can be used in the `CollectionsTypeTracking` module.
 * (Thereby avoiding naming conflicts with a "cousin" `AdditionalFlowStep` implementation.)
 */
abstract private class CollectionFlowStep extends DataFlow::AdditionalFlowStep {
  final override predicate step(DataFlow::Node pred, DataFlow::Node succ) { none() }

  final override predicate step(
    DataFlow::Node p, DataFlow::Node s, DataFlow::FlowLabel pl, DataFlow::FlowLabel sl
  ) {
    none()
  }

  /**
   * Holds if the property `prop` of the object `pred` should be loaded into `succ`.
   */
  predicate load(DataFlow::Node pred, DataFlow::Node succ, string prop) { none() }

  final override predicate loadStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
    this.load(pred, succ, prop)
  }

  /**
   * Holds if `pred` should be stored in the object `succ` under the property `prop`.
   */
  predicate store(DataFlow::Node pred, DataFlow::Node succ, string prop) { none() }

  final override predicate storeStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
    this.store(pred, succ, prop)
  }

  /**
   * Holds if the property `prop` should be copied from the object `pred` to the object `succ`.
   */
  predicate loadStore(DataFlow::Node pred, DataFlow::Node succ, string prop) { none() }

  final override predicate loadStoreStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
    this.loadStore(pred, succ, prop, prop)
  }

  /**
   * Holds if the property `loadProp` should be copied from the object `pred` to the property `storeProp` of object `succ`.
   */
  predicate loadStore(DataFlow::Node pred, DataFlow::Node succ, string loadProp, string storeProp) {
    none()
  }

  final override predicate loadStoreStep(
    DataFlow::Node pred, DataFlow::Node succ, string loadProp, string storeProp
  ) {
    this.loadStore(pred, succ, loadProp, storeProp)
  }

  /**
   * Holds if this step on a collection can load a value with a known key.
   */
  predicate canLoadValueWithKnownKey() { none() }
}

/**
 * A collection of predicates and clases for type-tracking collections.
 */
module CollectionsTypeTracking {
  /**
   * Gets the result from a single step through a collection, from `pred` to `result` summarized by `summary`.
   */
  pragma[inline]
  DataFlow::SourceNode collectionStep(DataFlow::Node pred, StepSummary summary) {
    exists(CollectionFlowStep step, string field |
      summary = LoadStep(field) and
      step.load(pred, result, field) and
      not (
        step.canLoadValueWithKnownKey() and // for a step that could load a known key,
        field = mapValueUnknownKey() // don't load values with an unknown key.
      )
      or
      summary = StoreStep(field) and
      step.store(pred, result, field)
      or
      summary = CopyStep(field) and
      step.loadStore(pred, result, field)
      or
      exists(string toField | summary = LoadStoreStep(field, toField) |
        step.loadStore(pred, result, field, toField)
      )
    )
  }

  /**
   * Gets the result from a single step through a collection, from `pred` with tracker `t2` to `result` with tracker `t`.
   */
  pragma[inline]
  DataFlow::SourceNode collectionStep(
    DataFlow::SourceNode pred, DataFlow::TypeTracker t, DataFlow::TypeTracker t2
  ) {
    exists(DataFlow::Node mid, StepSummary summary | pred.flowsTo(mid) and t = t2.append(summary) |
      result = collectionStep(mid, summary)
    )
  }

  /**
   * A class enabling the use of the collection related pseudo-properties in type-tracking predicates.
   */
  private class MapRelatedPseudoFieldAsTypeTrackingProperty extends TypeTrackingPseudoProperty {
    MapRelatedPseudoFieldAsTypeTrackingProperty() {
      this = [setElement(), iteratorElement()] or
      any(CollectionFlowStep step).store(_, _, this)
    }

    override string getLoadStoreToProp() {
      exists(CollectionFlowStep step | step.loadStore(_, _, this, result))
    }
  }
}

/**
 * A module for data-flow steps related standard library collection implementations.
 */
private module CollectionDataFlow {
  /**
   * A step for `Set.add()` method, which adds an element to a Set.
   */
  private class SetAdd extends CollectionFlowStep, DataFlow::MethodCallNode {
    SetAdd() { this.getMethodName() = "add" }

    override predicate store(DataFlow::Node element, DataFlow::Node obj, string prop) {
      obj = this.getReceiver().getALocalSource() and
      element = this.getArgument(0) and
      prop = setElement()
    }
  }

  /**
   * A step for the `Set` constructor, which copies any elements from the first argument into the resulting set.
   */
  private class SetConstructor extends CollectionFlowStep, DataFlow::NewNode {
    SetConstructor() { this = DataFlow::globalVarRef("Set").getAnInstantiation() }

    override predicate loadStore(
      DataFlow::Node pred, DataFlow::Node succ, string fromProp, string toProp
    ) {
      pred = this.getArgument(0) and
      succ = this and
      fromProp = arrayLikeElement() and
      toProp = setElement()
    }
  }

  /**
   * A step for a `for of` statement on a Map, Set, or Iterator.
   * For Sets and iterators the l-value are the elements of the set/iterator.
   * For Maps the l-value is a tuple containing a key and a value.
   *
   * This is partially duplicated behavior with the `for of` step for Arrays (in Arrays.qll).
   * This duplication is required for the type-tracking steps defined in `CollectionsTypeTracking`.
   */
  private class ForOfStep extends CollectionFlowStep, DataFlow::ValueNode {
    ForOfStmt forOf;
    DataFlow::Node element;

    ForOfStep() {
      this.asExpr() = forOf.getIterationDomain() and
      element = DataFlow::lvalueNode(forOf.getLValue())
    }

    override predicate load(DataFlow::Node obj, DataFlow::Node e, string prop) {
      obj = this and
      e = element and
      prop = arrayLikeElement()
    }

    override predicate loadStore(
      DataFlow::Node pred, DataFlow::Node succ, string fromProp, string toProp
    ) {
      pred = this and
      succ = element and
      fromProp = mapValueUnknownKey() and
      toProp = "1"
    }
  }

  /**
   * A step for a call to `forEach` on a Set or Map.
   */
  private class SetMapForEach extends CollectionFlowStep, DataFlow::MethodCallNode {
    SetMapForEach() { this.getMethodName() = "forEach" }

    override predicate load(DataFlow::Node obj, DataFlow::Node element, string prop) {
      obj = this.getReceiver() and
      element = this.getCallback(0).getParameter(0) and
      prop = [setElement(), mapValueUnknownKey()]
    }
  }

  /**
   * A call to the `get` method on a Map.
   * If the key of the call to `get` has a known string value, then only the value corresponding to that key will be retrieved.
   */
  private class MapGet extends CollectionFlowStep, DataFlow::MethodCallNode {
    MapGet() { this.getMethodName() = "get" }

    override predicate load(DataFlow::Node obj, DataFlow::Node element, string prop) {
      obj = this.getReceiver() and
      element = this and
      prop = mapValue(this.getArgument(0))
    }

    override predicate canLoadValueWithKnownKey() { any() }
  }

  /**
   * A call to the `set` method on a Map.
   *
   * If the key of the call to `set` has a known string value,
   * then the value will be saved into a pseudo-property corresponding to the known string value.
   * The value will additionally be saved into a pseudo-property corresponding to values with unknown keys.
   */
  private class MapSet extends CollectionFlowStep, DataFlow::MethodCallNode {
    MapSet() { this.getMethodName() = "set" }

    override predicate store(DataFlow::Node element, DataFlow::Node obj, string prop) {
      obj = this.getReceiver().getALocalSource() and
      element = this.getArgument(1) and
      // Makes sure that both known and unknown gets will work.
      prop = [mapValue(this.getArgument(0)), mapValueUnknownKey()]
    }
  }

  /**
   * A step for a call to `values` on a Map or a Set.
   */
  private class MapAndSetValues extends CollectionFlowStep, DataFlow::MethodCallNode {
    MapAndSetValues() { this.getMethodName() = "values" }

    override predicate loadStore(
      DataFlow::Node pred, DataFlow::Node succ, string fromProp, string toProp
    ) {
      pred = this.getReceiver() and
      succ = this and
      fromProp = [mapValueUnknownKey(), setElement()] and
      toProp = iteratorElement()
    }
  }

  /**
   * A step for a call to `keys` on a Set.
   */
  private class SetKeys extends CollectionFlowStep, DataFlow::MethodCallNode {
    SetKeys() { this.getMethodName() = "keys" }

    override predicate loadStore(
      DataFlow::Node pred, DataFlow::Node succ, string fromProp, string toProp
    ) {
      pred = this.getReceiver() and
      succ = this and
      fromProp = setElement() and
      toProp = iteratorElement()
    }
  }
}
