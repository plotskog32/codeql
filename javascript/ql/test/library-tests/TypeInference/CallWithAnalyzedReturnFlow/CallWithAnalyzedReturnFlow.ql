import javascript

from CallWithAnalyzedReturnFlow call
select call, call.getAValue()