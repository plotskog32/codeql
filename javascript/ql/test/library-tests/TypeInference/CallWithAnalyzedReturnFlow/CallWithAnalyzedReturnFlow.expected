| tst.js:4:1:4:16 | (function(){})() | file://:0:0:0:0 | undefined |
| tst.js:5:1:5:25 | (functi ... n; })() | file://:0:0:0:0 | undefined |
| tst.js:6:1:6:30 | (functi ... e; })() | file://:0:0:0:0 | true |
| tst.js:7:1:7:57 | (functi ... e; })() | file://:0:0:0:0 | false |
| tst.js:7:1:7:57 | (functi ... e; })() | file://:0:0:0:0 | true |
| tst.js:8:1:8:33 | (functi ... n; })() | file://:0:0:0:0 | indefinite value (global) |
| tst.js:11:5:11:8 | f1() | file://:0:0:0:0 | undefined |
| tst.js:23:5:23:8 | f4() | tst.js:21:16:21:30 | function f5 |
| tst.js:30:5:30:8 | f6() | tst.js:26:16:28:9 | function f7 |
